### PR TITLE
Saving images from simulator

### PR DESCRIPTION
**Pull request by Verichev**

### CODE REVIEW

**Review by bessszilard — COMMENTED** [review]
The car detects the traffic lights most of the time properly.
I assigned timestamp for the bug:
02:40 [min:sec] - car stops at the red light than slowly accelerates. This is not necessarily bad, because when after 1 sec, the traffic light will be green.
05:00 [min:sec] - the light variable changes, but there is no traffic light.
06:14 [min:sec] - car goes off the road, and loses the control.

File: ros/src/tl_detector/tl_detector.py [hunk 1]
```diff
@@ -11,11 +14,8 @@
 
 from light_classification.tl_classifier import TLClassifier
 
-# from darknet_ros_msgs.msg import BoundingBox
-# from darknet_ros_msgs.msg import BoundingBoxes
-
-
 STATE_COUNT_THRESHOLD = 3
+SAVING_IMAGES = True
```
Comment (alikureishy) [inline ros/src/tl_detector/tl_detector.py:18]: Let's keep this as a ROS param, so we needn't recompile whenever we need to change this value.

File: ros/src/tl_detector/tl_detector.py [hunk 2]
```diff
@@ -128,9 +133,25 @@ def get_light_state(self, light):
             self.prev_light_loc = None
             return False
         cv_image = self.bridge.imgmsg_to_cv2(self.camera_image, "bgr8")
-        return self.light_classifier.get_classification(cv_image)
+        if not SAVING_IMAGES:
+            traffic_color = self.light_classifier.get_classification(cv_image)
+            return traffic_color
+        else:
+            self.image_counter += 1
+            save_file = "../../../images/{}-{}.jpeg".format(self.traffic_color_to_file_name(light.state),
```
Comment (alikureishy) [inline ros/src/tl_detector/tl_detector.py:141]: Same comment as above ... we can make this a ros param and read it at the start.

File: ros/src/tl_detector/tl_detector.py [hunk 3]
```diff
@@ -34,9 +34,14 @@ def __init__(self):
...
         self.last_wp = -1
         self.state_count = 0
+        self.image_counter = 0
 
         self.bridge = CvBridge()
-        self.light_classifier = TLClassifier("frozen_inference_graph.pb")
+        if not SAVING_IMAGES:
+            self.light_classifier = TLClassifier("frozen_inference_graph.pb")
+        else:
+            if not os.path.exists("../../../images/"):
```
Comment (alikureishy) [inline ros/src/tl_detector/tl_detector.py:43]: Let's keep the _folder_ location ("../../../images/") as a config param...so we don't need to recompile, and we have one place to control these values.

**Review by alikureishy — DISMISSED** [review]

**Comment by alikureishy** [comment]
> The car detects the traffic lights most of the time properly.
> I assigned timestamp for the bug:
> 02:40 [min:sec] - car stops at the red light than slowly accelerates. This is not necessarily bad, because when after 1 sec, the traffic light will be green.
> 05:00 [min:sec] - the light variable changes, but there is no traffic light.
> 06:14 [min:sec] - car goes off the road, and loses the control.

This is probably because the io bottleneck is creating a race condition. As I had mentioned earlier, another node should be writing to disk, otherwise tl_detector has too much going on. It will also help with debugging if we separate the responsibilities of nodes this way.

**Comment by alikureishy** [comment]
I'm approving this one for check-in, but am working on creating a data collector node for actually writing the data. Meanwhile, you guys can continue with training.

**Review by alikureishy — APPROVED** [review]
I'm going to work on these 2 issues:
- #1 - Enable tl_detector to publish labeled data for traffic-light state classification to a new topic
- #3 - Create ROS node (tl_data_collector.py, for example) to subscribe to labeled data topic and save to disk

I will save the files to disk in the same way (same file naming format, same location) as the tl_detector.py is doing it right now, so that the rest of the pipeline (training etc) won't have to change. I will submit a review request when it's ready for integration.

**Comment by Verichev** [comment]
@bessszilard Check the green path, if it slowes down, the car will not catch up with planned path and go off the road.
Even without saving I have sometime this problem. This only work good on @safdark computer, cause he has raw linux. If you use this branch for saving images, use manual driving, it works perfectly. Than use images for feeding the model

**Comment by bessszilard** [comment]
You are right @Verichev . Probably the issue was with my machine. When the green path slowed down, I disabled the camera, and immediately the green path speed up, and continued everything as it have to be.

I am using Ubuntu 16.04 LTS on my notebook with kinetic 1.12.14.

**Pull request merged**